### ISSUE TITLE
How to use buildRustCrate?

### ISSUE THREAD

**Issue by matthiasbeyer** [NixOS/nixpkgs#32720]
## Issue description

From [this post](https://pijul.org/2017/12/12/buildrustcrate/) I read that there is a thing called `buildRustCrate` in nixpkgs now. The post links [this file](https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/rust/build-rust-crate.nix) as a resource. But I am not much more knowledgeable after reading it.

**How to use buildRustCrate**? There's documentation missing (at least I did not find any on nixos.org).

And, to include this somewhere: Is this build on `carnix`? Because I just used carnix to create a `build.nix` file for my rust project and it errors with `"Nix daemon out of memory"` after several seconds evaluating (on a 16GB RAM machine).

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.2, NixOS, 18.03pre122666.afe9649210c (Impala)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 1.11.16`
 - channels(root): `"nixos-18.03pre122666.afe9649210c"`
 - channels(m): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs`

**Comment by ne-sted** [comment]
Could you please post link to your `Cargo.lock`?

Did you see this file https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/rust.md#compiling-rust-crates-using-nix-instead-of-cargo ?

**Comment by matthiasbeyer** [comment]
@ne-sted my `Cargo.lock` is not yet checked into the source repository of the project, as the project is not in a usable state... so here is a pastebin: https://pastebin.com/SfAbetK2

I saw this link, yes. I did what's described there,... so my issue boils down to the second part from the initial message: 

> Because I just used carnix to create a build.nix file for my rust project and it errors with "Nix daemon out of memory" after several seconds evaluating (on a 16GB RAM machine).

Shall I edit the initial message and title of this issue?

**Comment by ne-sted** [comment]
@matthiasbeyer i don't really know how `cargo` work (i.e. workspaces). But for example i just cloned your https://github.com/matthiasbeyer/imag, generate Cargo.lock and run `carnix` on top of it.
Got this: https://gist.github.com/ne-sted/e7e2aa6d7ecdee4e8c6624b9554826d4

**Comment by matthiasbeyer** [comment]
@ne-sted yeah... now try to `nix-build <your file> -A imag_0_5_0` and wait for the out-of-memory error I posted above.

**Comment by ne-sted** [comment]
@matthiasbeyer just regenerated with `carnix --standalone` and... `nix-build <your file> -A imag_0_5_0` works fine! 8GB RAM there.

**Comment by matthiasbeyer** [comment]
O_o... okay... on my machine it is out of memory. 16GB RAM machine. Hm. I don't know what to do now.

**Comment by dpc** [comment]
@matthiasbeyer 

You might be recursively importing the whole project (with artefacts in `./target` etc.) which doesn't work well with `nix`. You can see `filters` I use in https://github.com/NixOS/nixpkgs/issues/31172

But for `carnix`, you could try wiping your `target` first or something.

**Comment by vitiral** [comment]
@matthiasbeyer just want to make sure, you're not using ramfs anywhere are you? I'm not sure where nix-rust stores temporary artifacts, but if it's /tmp and /tmp uses ramfs that could easily exhaust your memory.

**Comment by matthiasbeyer** [comment]
@vitiral I have `boot.tmpOnTmpfs = true` in my `configuration.nix` - so that could be the issue, yes.
